File: c/p5.py
# lcm.py - Project Euler problem 5, find LCM of factors 2-20
# Blake Mitchell, 2011


def isprime(n):
        if n < 2:
                return False
        i = 2 
        while i <= n // i:
                if n / i == float(n // i):
                        return False
                i += 1
        return True


def pdict_hit(pd, n):
        try:
                pd[n] += 1
        except KeyError:
                pd[n] = 1


def pdict_hit_greatest(pd, n, c):
        try:
                if (c > pd[n]):
                        pd[n] = c
        except KeyError:
                pd[n] = c


def pdict_product(pd):
        prod = 1
        for prime, count in pd.items():
                prod = prod * prime**count
        return prod


def primefac(n):
        pdict = {}
        if (isprime(n)):
                pdict_hit(pdict, n)
        else:
                i = 2
                while i <= n:
                        if isprime(i) and n / i == float(n // i):
                                pdict_hit(pdict, i)
                                n = n // i
                                i = 2
                        else:
                                i += 1
        return pdict


def make_lcm_primefac(pdlist):
        lcmpdict = {}
        for pdict in pdlist:
                for prime, count in pdict.items():
                        pdict_hit_greatest(lcmpdict, prime, count)
        return lcmpdict


def lcmrange(i):
        pdlist = [primefac(x) for x in range(2, i+1)]
        pdict = make_lcm_primefac(pdlist)
        return pdict_product(pdict)


if __name__ == '__main__':
        print('LCM of all factors between 2 and 20: ' + str(lcmrange(20)))

File: go/p5.go
package main

import "fmt"


type Plist map[int]int


func main() {
        fmt.Println(lcm_range(20))
}


func isprime(n int) bool {
        if n < 2 {
                return false
        }
        if n == 2 {
                return true
        }
        for i := 2; i <= n / i; i++ {
                if  n % i == 0 {
                        return false
                }
        }
        return true
}


func (pl *Plist) hit_greater(p, c int) {
        if c > (*pl)[p]  {
                (*pl)[p] = c
        }
}


func product(pl Plist) int {
        if len(pl) == 0 {
                return 0
        }
        prod := 1;
        for p, c := range pl {
                for i := 0; i < c; i++ {
                        prod *= p
                }
        }
        return prod
}


func (pl *Plist) primefac(n int) {
        if isprime(n) {
                (*pl)[n] = 1
                return
        }
        for i := 2; i <= n; i++ {
                if  isprime(i) && n % i == 0 {
                        (*pl)[i] = (*pl)[i] + 1
                        n = n / i
                        i = 1
                }
        }
}


func lcm_range(n int) int {
        lcmpl := make(Plist)
        for i := 2; i <= n; i++ {
                pl := make(Plist)
                pl.primefac(i)
                for p, c := range pl {
                        lcmpl.hit_greater(p, c)
                }
        }
        return product(lcmpl)
}
